src/spring-cloud/azext_spring_cloud/_validators.py

```diff
@@ -1,438 +1,550 @@
+                '{0} doesn\'t has valid CIDR prefix. '
+                ' --reserved-cidr-range should be 3 unused /16 IP ranges.'.format(ip))
+        return ip_address
+    except ValueError:
+        raise InvalidArgumentValueError('{0} is not a valid CIDR'.format(ip))
+
+
+def validate_vnet_required_parameters(namespace):
+    # pylint: disable=too-many-boolean-expressions
+    if not namespace.app_subnet and \
+       not namespace.service_runtime_subnet and \
+       not namespace.app_network_resource_group and \
+       not namespace.service_runtime_network_resource_group and \
+       not namespace.reserved_cidr_range and \
+       not namespace.vnet:
+        return
+    if namespace.sku and namespace.sku.lower() == 'basic':
+        raise InvalidArgumentValueError('Virtual Network Injection is not supported for Basic tier.')
+    if not namespace.app_subnet \
+       or not namespace.service_runtime_subnet:
+        raise InvalidArgumentValueError(
+            '--app-subnet, --service-runtime-subnet must be set when deploying to VNet')
+
+
+def validate_node_resource_group(namespace):
+    validate_vnet_required_parameters(namespace)
+    _validate_resource_group_name(namespace.service_runtime_network_resource_group,
+                                  'service-runtime-network-resource-group')
+    _validate_resource_group_name(namespace.app_network_resource_group, 'app-network-resource-group')
+
+
+def _validate_resource_group_name(name, message_name):
+    if not name:
+        return
+    matchObj = match(r'^[-\w\._\(\)]+$', name)
+    if matchObj is None:
+        raise InvalidArgumentValueError('--{0} must conform to the following pattern: \'^[-\\w\\._\\(\\)]+$\'.'.format(message_name))
+
+
+def _validate_route_table(namespace, vnet_obj):
+    app_route_table_id = ""
+    runtime_route_table_id = ""
+    for subnet in vnet_obj.subnets:
+        if subnet.id.lower() == namespace.app_subnet.lower() and subnet.route_table:
+            app_route_table_id = subnet.route_table.id
+        if subnet.id.lower() == namespace.service_runtime_subnet.lower() and subnet.route_table:
+            runtime_route_table_id = subnet.route_table.id
+
+    if app_route_table_id and runtime_route_table_id:
+        if app_route_table_id == runtime_route_table_id:
+            raise InvalidArgumentValueError('--service-runtime-subnet and --app-subnet should associate with different route tables.')
+    if (app_route_table_id and not runtime_route_table_id) \
+            or (not app_route_table_id and runtime_route_table_id):
+        raise InvalidArgumentValueError(
+            '--service-runtime-subnet and --app-subnet should both associate with different route tables or neither.')
+
+
+def validate_jar(namespace):
+    if namespace.disable_validation:
+        telemetry.set_user_fault("jar validation is disabled")
+        return
+    file_type, file_path = _get_upload_local_file(namespace.runtime_version, namespace.artifact_path)
```

Comment:
`_get_upload_local_file` packs the folder into a zip file. If the `artifact_path` is not set, the folder will be packed twice. It definitely does a more poor performance. And I am not sure whether it will throw exception at the second time to pack the folder for file already exists.

Reply:
fixed